ggml-cuda.cu

```diff
@@ -1960,20 +1960,84 @@ static void ggml_cuda_mul_mat(ggml_backend_cuda_context & ctx, const ggml_tensor
     }
 }
 
+struct mmid_row_mapping {
+    int64_t i1;
+    int64_t i2;
+};
+
+static __global__ void k_copy_src1_to_contiguous(const char * src1_original, char * src1_contiguous,
+                                                 int * cur_src1_row, mmid_row_mapping * row_mapping,
+                                                 const char * ids_dev, int64_t i02, int64_t ids_nb1, int64_t ids_nb0,
+                                                 int64_t ids_ne1, int64_t n_ids,
+                                                 int64_t ne11,
+                                                 size_t nb11, size_t nb12) {
+    int64_t iid1 = blockIdx.x;
+    int64_t id = blockIdx.y;
+
+    if (iid1 >= ids_ne1 || id >= n_ids) {
+        return;
+    }
+
+    const int32_t row_id_i = *(const int32_t *) (ids_dev + iid1*ids_nb1 + id*ids_nb0);
+
+    if (row_id_i != i02) {
+        return;
+    }
+
+    const int64_t i11 = id % ne11;
+    const int64_t i12 = iid1;
+
+    __shared__ int src1_row;
+    if (threadIdx.x == 0) {
+        src1_row = atomicAdd(cur_src1_row, 1);
+        row_mapping[src1_row] = {id, iid1};
+    }
+    __syncthreads();
+
+    const char * src1_row_original = src1_original + i11*nb11 + i12*nb12;
+    char * src1_row_contiguous = src1_contiguous + src1_row*nb11;
+
+    for (int i = threadIdx.x; i < nb11; i += blockDim.x) {
+        src1_row_contiguous[i] = src1_row_original[i];
```

Comment:
The copy would be faster if each thread were to copy 4 bytes at a time instead of a single byte.